src/dcrypto/v1/v1.go

```diff
@@ -91,19 +99,23 @@ func keys(pass, salt []byte, iterations int) (aesKey, hmacKey []byte, err error)
 	return aesKey, hmacKey, nil
 }
 
+// hashReadWriter hashes on write and on read finalizes the hash and returns it.
+// Writes after a Read will return an error.
 type hashReadWriter struct {
 	hash hash.Hash
 	done bool
 	sum  io.Reader
 }
 
+// Write implements io.Writer
 func (h *hashReadWriter) Write(p []byte) (int, error) {
 	if h.done {
 		return 0, errors.New("writing to hashReadWriter after read is not allowed")
 	}
 	return h.hash.Write(p)
 }
 
+// Read implements io.Reader.
 func (h *hashReadWriter) Read(p []byte) (int, error) {
 	if !h.done {
```

Comment:
I just noticed, don't we need mutual exclusivity protection if we are using `h.done` and setting `h.done`?